### PR TITLE
Data broker: extract permissions check

### PR DESCRIPTION
**Pull request by kael89**
Here we continue the cleanup that started in https://github.com/beyondessential/tupaia/pull/4698 by moving some abstract-data-source logic inline.

I recommend reviewing this one commit at a time - the commit message will indicate what each step is about.

### CODE REVIEW

File: packages/data-broker/src/DataBroker/checkPermissions.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ * Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+
+import { AccessPolicy } from '@tupaia/access-policy';
+import { DataBrokerModelRegistry, DataElement, DataSource } from '../types';
+import { fetchOrgUnitsByCountry } from './fetchOrgUnitsByCountry';
+
+const BES_ADMIN_PERMISSION_GROUP = 'BES Admin';
+
+const getUserPermissions = (accessPolicy?: AccessPolicy, countryCodes?: string[]) => {
```
Comment (kael89) [inline packages/data-broker/src/DataBroker/checkPermissions.ts]: This diff is purely moving those functions from `DataBroker`, and changing the function signatures to accept all required input

File: packages/data-broker/src/DataBroker/DataBroker.ts [hunk 2]
```diff
@@ -246,17 +236,15 @@ export class DataBroker {
     const syncGroups = await fetchSyncGroups(this.models, syncGroupCodes);
     const organisationUnitCodes = getOrganisationUnitCodes(options);
     const pulls = await this.getPulls(syncGroups, organisationUnitCodes);
-    const type = this.getDataSourceTypes().SYNC_GROUP;
```
Comment (kael89) [inline packages/data-broker/src/DataBroker/DataBroker.ts]: No permissions check for sync group

File: packages/data-broker/src/DataBroker/DataBroker.ts [hunk 3]
```diff
@@ -269,30 +257,6 @@ export class DataBroker {
     );
   }
 
-  private pullForServiceAndType = async (
```
Comment (kael89) [inline packages/data-broker/src/DataBroker/DataBroker.ts:382]: This is inlined now

File: packages/data-broker/src/DataBroker/DataBroker.ts [hunk 4]
```diff
@@ -306,19 +178,23 @@ export class DataBroker {
...
     const nestedResults = await Promise.all(
-      pulls.map(({ dataSources: dataSourcesForThisPull, serviceType, dataServiceMapping }) => {
-        return this.pullForServiceAndType(
-          dataSourcesForThisPull,
-          validatedOptions,
-          type,
-          serviceType,
+      pulls.map(({ dataSources: dataElementsForThisPull, serviceType, dataServiceMapping }) => {
+        const service = this.createService(serviceType);
+        return service.pull(dataElementsForThisPull, this.getDataSourceTypes().DATA_ELEMENT, {
```
Comment (kael89) [inline packages/data-broker/src/DataBroker/DataBroker.ts:193]: The refactoring of this PR will allow us to change the `Service` interface to something like the following:

```
return service.pullAnalytics(dataElementsForThisPull, options);
```

File: packages/data-broker/src/DataBroker/checkPermissions.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      throw new Error(`Missing permissions to the following data elements: ${missingPermissions}`);
+    }
+
+    return organisationUnitCodes;
+  }
+
+  const organisationUnitsByCountry = await fetchOrgUnitsByCountry(models, organisationUnitCodes);
+  const countryCodes = Object.keys(organisationUnitsByCountry);
+
+  let organisationUnitsWithPermission: string[] = [];
```
Comment (edmofro) [inline packages/data-broker/src/DataBroker/checkPermissions.ts]: Lost some of the nice renames in moving these out

File: packages/data-broker/src/DataBroker/getAllowedOrgUnits.ts [hunk 6]
```diff
@@ -35,19 +35,19 @@ export const checkDataElementPermissions = async (
...
+    return orgUnitCodes;
   }
 
-  const organisationUnitsByCountry = await fetchOrgUnitsByCountry(models, organisationUnitCodes);
-  const countryCodes = Object.keys(organisationUnitsByCountry);
+  const orgUnitsByCountry = await fetchOrgUnitsByCountry(models, orgUnitCodes);
+  const countryCodes = Object.keys(orgUnitsByCountry);
 
-  let organisationUnitsWithPermission: string[] = [];
+  let allowedOrgUnits: string[] = [];
```
Comment (edmofro) [inline packages/data-broker/src/DataBroker/getAllowedOrgUnits.ts]: Regained nice renames

File: packages/data-broker/src/DataBroker/fetchAllowedOrgUnits.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@
  */
 
 import { AccessPolicy } from '@tupaia/access-policy';
-import { DataBrokerModelRegistry, DataElement, DataGroup, DataSource } from '../types';
```
Comment (edmofro) [inline packages/data-broker/src/DataBroker/fetchAllowedOrgUnits.ts]: Good lint catch at the same time

**Review by edmofro — APPROVED** [review]
Thanks again @kael89, great call to inline those permission checks! Really well laid out PR too, reviewing commit by commit felt like I was watching a live demo of refactoring techniques ;-) 

I'm happy with all of the changes here, and it's a good day to merge as we have just done a Tupaia release so this will have 2 weeks of passive regression before the next release

**Comment by kael89** [comment]
> Thanks again @kael89, great call to inline those permission checks! Really well laid out PR too, reviewing commit by commit felt like I was watching a live demo of refactoring techniques ;-)
> 
> I'm happy with all of the changes here, and it's a good day to merge as we have just done a Tupaia release so this will have 2 weeks of passive regression before the next release

Awesome! I'll merge now, passive regression testing strategy sounds great.

As always, let me know of any bugs with those PRs and I'd be happy to fix them asap - the only caveat being I'll only be available after hours.

**Pull request merged**